### PR TITLE
System-wide webhooks

### PR DESCRIPTION
**Pull request by jamesorlakin**
#10545 was indeed too basic. This PR attempts to add webhooks that work system-wide, regardless of repository. The intent is to allow a savvy sysadmin to extend Gitea's functionality by being able to listen for all events - opened PRs/comments to add Jira cards to, deleting Docker images for the repo, etc. This also leaves default webhooks intact with a lot of code reuse but I fear it may be a tad confusing.

Tasks:
- [X] Create a new column to differentiate system webhooks from default ones
- [X] Adjust the router to serve both `hooks` and `system-hooks` reusing the same handlers
- [X] Tweak the creation of webhooks to determine if setting a value in the new column is necessary
- [x] Create modifications to the template and locales to explain the differences between the two
- [x] Fetch system webhooks and make use of them during events
- [ ] Add API endpoints (will attempt in a separate PR)
...and probably something else

I'm more than happy to receive any feedback. This is my biggest PR so far but I'm growing quite fond of Go! 🚀 

image

### CODE REVIEW

File: models/webhook.go [hunk 1]
```diff
@@ -315,7 +316,7 @@ func GetWebhooksByOrgID(orgID int64, listOptions ListOptions) ([]*Webhook, error
 func GetDefaultWebhook(id int64) (*Webhook, error) {
 	webhook := &Webhook{ID: id}
 	has, err := x.
-		Where("repo_id=? AND org_id=?", 0, 0).
+		Where("repo_id=? AND org_id=? AND is_system_webhook=?", 0, 0, 0).
```
Comment (lafriks) [inline models/webhook.go]:
```suggestion
		Where("repo_id=? AND org_id=? AND is_system_webhook=?", 0, 0, false).
```

File: models/webhook.go [hunk 2]
```diff
@@ -333,7 +334,33 @@ func GetDefaultWebhooks() ([]*Webhook, error) {
 func getDefaultWebhooks(e Engine) ([]*Webhook, error) {
 	webhooks := make([]*Webhook, 0, 5)
 	return webhooks, e.
-		Where("repo_id=? AND org_id=?", 0, 0).
+		Where("repo_id=? AND org_id=? AND is_system_webhook=?", 0, 0, 0).
```
Comment (lafriks) [inline models/webhook.go]:
```suggestion
		Where("repo_id=? AND org_id=? AND is_system_webhook=?", 0, 0, false).
```

File: models/webhook.go [hunk 3]
```diff
@@ -333,7 +334,33 @@ func GetDefaultWebhooks() ([]*Webhook, error) {
...
+		Find(&webhooks)
+}
+
+// GetSystemWebhook returns admin system webhook by given ID.
+func GetSystemWebhook(id int64) (*Webhook, error) {
+	webhook := &Webhook{ID: id}
+	has, err := x.
+		Where("repo_id=? AND org_id=? AND is_system_webhook=?", 0, 0, 1).
```
Comment (lafriks) [inline models/webhook.go]:
```suggestion
		Where("repo_id=? AND org_id=? AND is_system_webhook=?", 0, 0, true).
```

File: models/webhook.go [hunk 4]
```diff
@@ -333,7 +334,33 @@ func GetDefaultWebhooks() ([]*Webhook, error) {
...
+
+// GetSystemWebhooks returns all admin system webhooks.
+func GetSystemWebhooks() ([]*Webhook, error) {
+	return getSystemWebhooks(x)
+}
+
+func getSystemWebhooks(e Engine) ([]*Webhook, error) {
+	webhooks := make([]*Webhook, 0, 5)
+	return webhooks, e.
+		Where("repo_id=? AND org_id=? AND is_system_webhook=?", 0, 0, 1).
```
Comment (lafriks) [inline models/webhook.go]:
```suggestion
		Where("repo_id=? AND org_id=? AND is_system_webhook=?", 0, 0, true).
```

File: models/webhook.go [hunk 5]
```diff
@@ -394,7 +421,7 @@ func DeleteDefaultWebhook(id int64) error {
 	}
 
 	count, err := sess.
-		Where("repo_id=? AND org_id=?", 0, 0).
+		Where("repo_id=? AND org_id=? AND is_system_webhook=?", 0, 0, 0).
```
Comment (lafriks) [inline models/webhook.go]:
```suggestion
		Where("repo_id=? AND org_id=? AND is_system_webhook=?", 0, 0, false).
```

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/migrations/v128.go [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+import (
+	"fmt"
+
+	"xorm.io/xorm"
+)
+
+func addSystemWebhookColumn(x *xorm.Engine) error {
+	type Webhook struct {
+		IsSystemWebhook bool `xorm:"NOT NULL DEFAULT 0"`
```
Comment (lafriks) [inline models/migrations/v128.go]:
```suggestion
		IsSystemWebhook bool `xorm:"NOT NULL DEFAULT false"`
```

**Comment by jamesorlakin** [comment]
@lafriks Oh yep! false and 0 should be different things... 👍

**Comment by jamesorlakin** [comment]
Looking at what's currently there, there's no API for default webhooks. Should I look at trying to add that for both default and system webhooks (maybe another PR) or just leave everything as-is?

**Comment by lafriks** [comment]
You can leave that for different PR. If code is too large it is harder to review and it takes more time to get it merged

File: templates/admin/nav.tmpl [hunk 7]
```diff
@@ -1,16 +0,0 @@
-<div id="setting-menu" class="grid-1-5 panel panel-radius left">
```
Comment (lafriks) [inline templates/admin/nav.tmpl]: Why this file was deleted?

Comment (jamesorlakin) [reply templates/admin/nav.tmpl]: I couldn't find any references to it in the code base, and no changes other than 1 for the last 4 years. I originally modified it before I realised `navbar` was the right one.

Comment (lafriks) [reply templates/admin/nav.tmpl]: Please open sepate PR for that

**Review by lafriks — APPROVED** [review]

File: routers/repo/webhook.go [hunk 8]
```diff
@@ -150,7 +165,7 @@ func ParseHookEvent(form auth.WebhookForm) *models.HookEvent {
 	}
 }
 
-// WebHooksNewPost response for creating webhook
+// WebHooksNewPost response for creating Gitea webhook
```
Comment (6543) [inline routers/repo/webhook.go]: Only gitea webhook?

**Review by 6543 — CHANGES_REQUESTED** [review]

File: routers/repo/webhook.go (hunk 8 above)
Comment (jamesorlakin) [reply routers/repo/webhook.go]: Yep, in `routes.go` only the new Gitea hook uses this handler:
```
			m.Get("/:type/new", repo.WebhooksNew)
			m.Post("/gitea/new", bindIgnErr(auth.NewWebhookForm{}), repo.WebHooksNewPost)
			m.Post("/gogs/new", bindIgnErr(auth.NewGogshookForm{}), repo.GogsHooksNewPost)
			m.Post("/slack/new", bindIgnErr(auth.NewSlackHookForm{}), repo.SlackHooksNewPost)
```

I suppose the method should be renamed really.

Comment (6543) [reply routers/repo/webhook.go]: Can you rename it?

**Comment by lunny** [comment]
Please resolve the conflicts

File: routers/repo/webhook.go (hunk 8 above)
Comment (jamesorlakin) [reply routers/repo/webhook.go]: Sure :+1:

**Review by 6543 — APPROVED** [review]

**Comment by lafriks** [comment]
Make lgtm work

**Comment by guillep2k** [comment]
What about adding a `changelog` label?

**Pull request merged**